pkg/entrypoint/admissioncontroller/podwhitelist.go

```diff
@@ -5,161 +5,168 @@ import (
-	case *extensions.ReplicaSet:
-		pod := &core.Pod{ObjectMeta: o.Spec.Template.ObjectMeta, Spec: o.Spec.Template.Spec}
-		pod.Namespace = o.Namespace
-		return pod, nil
-	case *apps.StatefulSet:
-		pod := &core.Pod{ObjectMeta: o.Spec.Template.ObjectMeta, Spec: o.Spec.Template.Spec}
-		pod.Namespace = o.Namespace
-		return pod, nil
-	case *batch.Job:
-		pod := &core.Pod{ObjectMeta: o.Spec.Template.ObjectMeta, Spec: o.Spec.Template.Spec}
-		pod.Namespace = o.Namespace
-		return pod, nil
-	case *batch.CronJob:
-		pod := &core.Pod{ObjectMeta: o.Spec.JobTemplate.Spec.Template.ObjectMeta, Spec: o.Spec.JobTemplate.Spec.Template.Spec}
-		pod.Namespace = o.Namespace
-		return pod, nil
-	case *oapps.DeploymentConfig:
-		pod := &core.Pod{ObjectMeta: o.Spec.Template.ObjectMeta, Spec: o.Spec.Template.Spec}
-		pod.Namespace = o.Namespace
-		return pod, nil
-	case *extensions.Deployment:
-		pod := &core.Pod{ObjectMeta: o.Spec.Template.ObjectMeta, Spec: o.Spec.Template.Spec}
-		pod.Namespace = o.Namespace
-		return pod, nil
-	default:
-		return nil, fmt.Errorf("unimplemented gvk %v", o.GetObjectKind().GroupVersionKind())
-	}
-}
-
 func (ac *admissionController) handleWhitelist(w http.ResponseWriter, r *http.Request) {
-	req, errcode := ac.getAdmissionReviewRequest(r)
+	req, errcode := ac.getAdmissionRequest(r)
 	if errcode != 0 {
 		http.Error(w, http.StatusText(errcode), errcode)
 		return
 	}
-	if req.UID == "" || req.Kind.Version == "" || req.Kind.Kind == "" {
-		http.Error(w, http.StatusText(http.StatusBadRequest), http.StatusBadRequest)
+
+	errs, err := ac.validateWhitelistRequest(req)
+	if err != nil {
+		ac.log.Error(err)
+		http.Error(w, http.StatusText(http.StatusInternalServerError), http.StatusInternalServerError)
 		return
 	}
-	//whitelist system namespaces
-	systemNamespaces := map[string]bool{
-		"default":                           true,
-		"kube-public":                       true,
-		"kube-service-catalog":              true,
-		"kube-system":                       true,
-		"openshift-ansible-service-broker":  true,
-		"openshift-azure":                   true,
-		"openshift-azure-branding":          true,
-		"openshift-azure-logging":           true,
-		"openshift-azure-monitoring":        true,
-		"openshift-console":                 true,
-		"openshift-etcd":                    true,
-		"openshift-infra":                   true,
-		"openshift-logging":                 true,
-		"openshift-monitoring":              true,
-		"openshift-node":                    true,
-		"openshift-sdn":                     true,
-		"openshift-template-service-broker": true,
-		"openshift-web-console":             true,
-	}
-	if _, present := systemNamespaces[req.Namespace]; present {
-		ac.sendResult(nil, w, req.UID)
-		return
+
+	ac.sendReview(w, req, errs.ToAggregate())
+}
+
+func (ac *admissionController) validateWhitelistRequest(req *admission.AdmissionRequest) (field.ErrorList, error) {
+	if strings.HasPrefix(req.Namespace, "openshift-") ||
```

Comment:
So we relay here on OCP admission controller to make sure users cant create `openshift-myproject` namespace where.  Is `os adm new-project` is not available to users currently?

Reply:
It should be the case that ARO users can't create *namespaces*, only *projects*.  The project request infrastructure protects against creation of projects with reserved names.